### PR TITLE
Orc: Closes #5777 - Obtain ORC stripe offsets from writer 

### LINKED ISSUE #5777: ORCFileAppender should fetch the stripe offsets from the Writer instead of opening the written file
### Feature Request / Improvement

Currently the ORCFileAppender opens a written ORC file just to derive the Stripe offsets

```java
  @Override
  public List<Long> splitOffsets() {
    Preconditions.checkState(isClosed, "File is not yet closed");
    try (Reader reader = ORC.newFileReader(file.toInputFile(), conf)) {
      List<StripeInformation> stripes = reader.getStripes();
      return Collections.unmodifiableList(Lists.transform(stripes, StripeInformation::getOffset));
    } catch (IOException e) {
      throw new RuntimeIOException(e, "Can't close ORC reader %s", file.location());
    }
  }
```
Starting with ORC 1.7 we have added a public API to retrieve offset information from the writer. This when called after close gives the complete stripe information that is written out to the file. With this we can avoid opening the written file.

### Query engine

_No response_

### PR DESCRIPTION
**Pull request by pavibhai**
## What?
Fixes ORCFileAppender such that it can retrieve the stripe offsets from the writer instead of having to reopen the file.

## Why?
From ORC 1.7 we have a public API for getting stripe information from the writer. This avoids an additional IO to get stripe offsets

## Tests
Added unit test to verify that the offsets are matching the offsets read from the file

### CODE REVIEW

**Comment by pavibhai** [comment]
@kbendick Hope you are still looking at ORC support on Iceberg. Can you please take a look at this PR?

File: orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java [hunk 1]
```diff
@@ -147,11 +146,12 @@ public long length() {
...
     Preconditions.checkState(isClosed, "File is not yet closed");
-    try (Reader reader = ORC.newFileReader(file.toInputFile(), conf)) {
-      List<StripeInformation> stripes = reader.getStripes();
+    try {
+      List<StripeInformation> stripes = writer.getStripes();
       return Collections.unmodifiableList(Lists.transform(stripes, StripeInformation::getOffset));
     } catch (IOException e) {
-      throw new RuntimeIOException(e, "Can't close ORC reader %s", file.location());
+      throw new RuntimeIOException(
+          e, "Cannot receive stripe information from writer for %s", file.location());
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java:154]: "Cannot receive" seems like a strange way to state that `getStripes` failed. What about `Failed to get stripe information from writer for: %s`?

File: orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+            .build()) {
+      writtenRecords = Lists.newArrayList(reader);
+    }
+
+    Assert.assertEquals("Written records should match", records, writtenRecords);
+  }
+
+  @SuppressWarnings("checkstyle:AvoidEscapedUnicodeCharacters")
+  @Test
+  public void testInvalidUpperBoundString() throws Exception {
```
Comment (rdblue) [inline orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:137]: Is this related to this PR? What is the purpose of this test?

File: orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java [hunk 3]
```diff
@@ -147,11 +146,12 @@ public long length() {
   @Override
   public List<Long> splitOffsets() {
     Preconditions.checkState(isClosed, "File is not yet closed");
-    try (Reader reader = ORC.newFileReader(file.toInputFile(), conf)) {
-      List<StripeInformation> stripes = reader.getStripes();
+    try {
+      List<StripeInformation> stripes = writer.getStripes();
```
Comment (rdblue) [inline orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java:150]: This change looks great to me. Thanks for updating this, @pavibhai!

**Review by rdblue — CHANGES_REQUESTED** [review]
Looks good other than an error message and some tests for unrelated cases.

File: orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java (hunk 1 above)
Comment (pavibhai) [reply orc/src/main/java/org/apache/iceberg/orc/OrcFileAppender.java:154]: Makes sense, will change.

File: orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java (hunk 2 above)
Comment (pavibhai) [reply orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:137]: I was noticed that there was no unit test on ORC write in the project. I ported the current parquet test and introduced the check on offsets.

If you would prefer that I raise a separate PR on other unit test cases then I can split it.

Comment (pavibhai) [reply orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:137]: Removing the additional unit tests, will raise a separate PR for them

File: orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+            .project(SCHEMA)
+            .createReaderFunc(fileSchema -> GenericOrcReader.buildReader(SCHEMA, fileSchema))
+            .build()) {
+      writtenRecords = Lists.newArrayList(reader);
+    }
+
+    Assert.assertEquals("Written records should match", overflowRecords, writtenRecords);
+
+    // Expect bounds on LONG column
+    Assert.assertTrue("Should have a valid lower bound", dataFile.lowerBounds().containsKey(1));
```
Comment (nastra) [inline orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:273]:
```suggestion
    Assertions.assertThat(dataFile.lowerBounds()).containsKey(1));
```

Comment (nastra) [reply orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:273]: this will show you the content of the map if the assertion ever fails, rather than just showing true != false.
would be great to apply this to the other places as well

Comment (pavibhai) [reply orc/src/test/java/org/apache/iceberg/orc/TestOrcDataWriter.java:273]: Thanks @nastra as suggested earlier I am removing the additional tests and will raise a separate PR. Will incorporate this feedback into that PR.

**Comment by pavibhai** [comment]
> Looks good other than an error message and some tests for unrelated cases.

@rdblue I have addressed the review comments can you please take another look at this?

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
Thanks, @pavibhai! Looks great so I merged it.
